### PR TITLE
Corrected ttb text positioning

### PR DESCRIPTION
**Pull request by radarhere**
Helps #2656 and #3125

libraqm 0.4 and 0.5 throw an error for ttb - see https://github.com/HOST-Oman/libraqm/commit/7088743be4afe22e82282912bf414f5051046bb8. 0.3 works, but since there is no simple way to test for the difference between 0.3 and 0.4, for simplicity I'm saying that ttb requires libraqm 0.6 or later.

### CODE REVIEW

**Comment by hugovk** [comment]
Thanks! 

Is there a way to get the libraqm version programatically? If not, shall we open an issue with libraqm to ask them to add some API? It may be useful in the future.

**Comment by radarhere** [comment]
To be clear, this is testing to see if libraqm 0.6 is present, just not through a version string - I do it by testing for the existence of an unrelated function that was added in 0.6.

I've created https://github.com/HOST-Oman/libraqm/issues/103 requesting a version identifier.

**Comment by hugovk** [comment]
Thanks! It could be that the unrelated feature is removed in 0.7 or some future version, and then breaking this fix.

File: src/_imagingft.c [hunk 1]
```diff
@@ -498,20 +506,20 @@ text_layout_raqm(PyObject* string, FontObject* self, const char* dir, PyObject *
         goto failed;
     }
 
-    if (p_raqm.version == 1){
-        for (i = 0; i < count; i++) {
+    for (i = 0; i < count; i++) {
+        if (p_raqm.version == 1) {
```
Comment (khaledhosny) [inline src/_imagingft.c]: I think it would be faster to check the condition outside the loop.

Comment (khaledhosny) [inline src/_imagingft.c]: I think it would be faster to check the condition outside the loop.

File: Tests/test_imagefontctl.py [hunk 2]
```diff
@@ -104,6 +104,22 @@ def test_text_direction_rtl2(self):
 
         self.assert_image_similar(im, target_img, .5)
 
+    def test_text_direction_ttb(self):
+        ttf = ImageFont.truetype(FONT_PATH, FONT_SIZE)
+
+        im = Image.new(mode='RGB', size=(100, 300))
+        draw = ImageDraw.Draw(im)
+        try:
+            draw.text((0, 0), 'English عربي', font=ttf, fill=500, direction='ttb')
```
Comment (khaledhosny) [inline Tests/test_imagefontctl.py]: Wouldn’t it be better to actually test some CJK character here?

File: src/_imagingft.c [hunk 3]
```diff
@@ -748,73 +797,80 @@ font_render(FontObject* self, PyObject* args)
...
                     }
                 }
-                source += glyph->bitmap.pitch;
             }
+            source += glyph->bitmap.pitch;
+        }
+        if (horizontal_dir) {
+            x += glyph_info[i].x_advance;
+        } else {
+            y -= glyph_info[i].y_advance;
```
Comment (khaledhosny) [inline src/_imagingft.c]: I think the condition is not needed, you can assign x and y unconditionally. In horizontal direction `y_advance` will be zero and in vertical direction `x_advance` will be zero.

File: src/_imagingft.c (hunk 1 above)
Comment (radarhere) [reply src/_imagingft.c]: You're right. I've updated the commit

File: src/_imagingft.c (hunk 3 above)
Comment (radarhere) [reply src/_imagingft.c]: Okay. I've updated the commit.

File: src/_imagingft.c (hunk 1 above)
Comment (radarhere) [reply src/_imagingft.c]: I'm having trouble finding a small CJK font that could be included within the Pillow license. Do you have any thoughts?

File: Tests/test_imagefontctl.py (hunk 2 above)
Comment (radarhere) [reply Tests/test_imagefontctl.py]: I'm having trouble finding a small CJK font that could be included within the Pillow license. Do you have any thoughts?

Comment (radarhere) [reply Tests/test_imagefontctl.py]: Actually, I've found one that's not too large.

Comment (khaledhosny) [reply Tests/test_imagefontctl.py]: You can also subset any font with `pyftsubset` (from [FontTools](https://github.com/fonttools/fonttools)) to include just the few characters used in the test.

File: src/_imagingft.c (hunk 1 above)
Comment (radarhere) [reply src/_imagingft.c]: This looks like a duplicate comment.

**Comment by hugovk** [comment]
Now https://github.com/HOST-Oman/libraqm/pull/104 is merged and released in https://github.com/HOST-Oman/libraqm/releases/tag/v0.7.0, can we replace the version inference with a direct check for 0.7.0+?

**Comment by radarhere** [comment]
Okay, done.

**Pull request merged**